apollo-normalized-cache/src/commonMain/kotlin/com/apollographql/apollo3/cache/normalized/ClientCacheExtensions.kt

```diff
@@ -232,36 +260,142 @@ internal val <D : Operation.Data> ApolloRequest<D>.cacheHeaders
 internal val <D : Operation.Data> ApolloRequest<D>.watch
   get() = executionContext[WatchContext]?.value ?: false
 
-
-class CacheInfo(
-    val millisStart: Long,
-    val millisEnd: Long,
-    val hit: Boolean,
-    val missedKey: String?,
-    val missedField: String?,
+/**
+ * @param cacheHit true if this was a cache hit
+ * @param cacheException the exception while reading the cache. Note that it's possible to have [cacheHit] == false && [cacheException] == null
+ * if no cache read was attempted
+ */
+class CacheInfo private constructor(
+    val cacheStartMillis: Long,
+    val cacheEndMillis: Long,
+    val networkStartMillis: Long,
+    val networkEndMillis: Long,
+    val cacheHit: Boolean,
+    val cacheMissException: CacheMissException?,
+    val networkException: ApolloException?,
 ) : ExecutionContext.Element {
+
+  @Deprecated("Use CacheInfo.Builder")
+  constructor(
+      millisStart: Long,
+      millisEnd: Long,
+      hit: Boolean,
+      missedKey: String?,
+      missedField: String?,
+  ) : this(
+      cacheStartMillis = millisStart,
+      cacheEndMillis = millisEnd,
+      networkStartMillis = 0,
+      networkEndMillis = 0,
+      cacheHit = hit,
+      cacheMissException = missedKey?.let { CacheMissException(it, missedField) },
+      networkException = null
+  )
+
   override val key: ExecutionContext.Key<*>
     get() = Key
 
+  @Deprecated("Use cacheStartMillis instead", ReplaceWith("cacheStartMillis"))
```

Comment:
A completely adjacent thought: could it be a good idea to keep a log of when (in which version) APIs were deprecated, if we need to reason about when they can be deleted. Kotlin has a `DeprecatedSinceKotlin` annotation. Maybe I'm overthinking and it's unnecessary though :)

Reply:
I'm always for more strict API versionning 👍

Reply:
Follow up issue: https://github.com/apollographql/apollo-kotlin/issues/3745